src/broker/content-mmap.c

```diff
@@ -0,0 +1,766 @@
+ * in missing cache entries.  Design tradeoff:  mapping and lookup are fast,
+ * and the cache implementation is lightweight and simple, at the expense of
+ * unmap efficiency.
+ */
+static int plug_cache_holes (struct content_mmap *mm)
+{
+    const char *name;
+    struct content_region *reg;
+
+    name = hola_hash_first (mm->tags);
+    while (name) {
+        reg = hola_list_first (mm->tags, name);
+        while (reg) {
+            if (region_cache_add (reg) < 0)
+                return -1;
+            reg = hola_list_next (mm->tags, name);
+        }
+        name = hola_hash_next (mm->tags);
+    }
+    return 0;
+}
+
+// zhashx_destructor_fn footprint
+static void cache_entry_destructor (void **item)
+{
+    if (item) {
+        cache_entry_destroy (*item);
+        *item = NULL;
+    }
+}
+// zhashx_hash_fn footprint
+static size_t cache_entry_hasher (const void *key)
+{
+    return *(size_t *)key;
+}
+// zhashx_comparator_fn footprint
+static int cache_entry_comparator (const void *item1, const void *item2)
+{
+    return memcmp (item1, item2, content_hash_size);
+}
+
+void content_mmap_region_decref (struct content_region *reg)
+{
+    if (reg && --reg->refcount == 0) {
+        int saved_errno = errno;
+        region_cache_remove (reg);
+        if (reg->data != MAP_FAILED)
+            (void)munmap (reg->data, reg->data_size);
+        json_decref (reg->fileref);
+        free (reg->fullpath);
+        free (reg->blobref);
+        free (reg->fileref_data);
+        free (reg);
+        errno = saved_errno;
+    }
+}
+
+// zhashx_destructor_fn footprint
+static void content_mmap_region_destructor (void **item)
+{
+    if (item) {
+        content_mmap_region_decref (*item);
+        *item = NULL;
+    }
+}
+
+struct content_region *content_mmap_region_incref (struct content_region *reg)
+{
+    if (reg)
+        reg->refcount++;
+    return reg;
+}
+
+/* Validate mmapped blob before use, checking for:
+ * - size has changed so mmmapped pages are no longer valid (SIGBUS if used!)
+ * - content no longer matches hash
+ * To avoid repeatedly calling stat(2) on a file, skip it if last check was
+ * within max_check_age seconds.
+ */
+bool content_mmap_validate (struct content_region *reg,
+                            const void *hash,
+                            int hash_size,
+                            const void *data,
+                            int data_size)
+{
+    char hash2[BLOBREF_MAX_DIGEST_SIZE];
+
+    // no need to check the fileref blob as it was not mmapped
+    if (data == reg->fileref_data)
+        return true;
+
+    assert (reg->data != NULL);
+    assert (data >= reg->data);
+    assert (data + data_size <= reg->data + reg->data_size);
+
+    if (monotime_since (reg->last_check)/1000 >= max_check_age) {
+        struct stat sb;
+
+        if (stat (reg->fullpath, &sb) < 0
+            || sb.st_size < reg->data_size)
+            return false;
```

Comment:
What problem were you thinking this would solve?  The stat already fails if we can no longer read the file.  Since we're taking a copy from a specific point in time, the main thing we need to protect against is SIGBUS errors (file shrinks) and violating the content addressable storage invariant that blobs never change IMHO.

Reply:
oh yeah, i'm stupid ... the `stat()` is effectively a perms check :P